src/Bicep.Core/LanguageConstants.cs

```diff
@@ -48,37 +48,32 @@ public static class LanguageConstants
 
         public static readonly string DeclarationTypesString = LanguageConstants.DeclarationTypes.Keys.ConcatString(ListSeparator);
 
-        public static TypeSymbol CreateParameterModifierType(TypeSymbol parameterType)
+        public static TypeSymbol CreateParameterModifierType(TypeSymbol primitiveType, TypeSymbol allowedValuesType)
         {
-            if (parameterType.TypeKind != TypeKind.Primitive)
-            {
-                throw new ArgumentException($"Modifiers are not supported for type '{parameterType.Name}'.");
-            }
-
-            return new NamedObjectType($"ParameterModifier_{parameterType.Name}", CreateParameterModifierProperties(parameterType), additionalPropertiesType: null);
+            return new NamedObjectType($"ParameterModifier<{allowedValuesType.Name}>", CreateParameterModifierProperties(primitiveType, allowedValuesType), additionalPropertiesType: null);
         }
 
-        private static IEnumerable<TypeProperty> CreateParameterModifierProperties(TypeSymbol parameterType)
+        private static IEnumerable<TypeProperty> CreateParameterModifierProperties(TypeSymbol primitiveType, TypeSymbol allowedValuesType)
```

Comment:
>TypeSymbol primitiveType, TypeSymbol allowedValuesType [](start = 83, length = 54)

Are these ever different types? In `TypeValidatorAssignabilityTests`, every invocation has the same type for both.

Reply:
Yeah, primitiveType can be "string" and allowedValuesType can be a string literal or union of string literals. I'll add a test for this.